### PR TITLE
Workaround for #507

### PR DESCRIPTION
**Pull request by cmyr**
This doesn't fully resolve the underlying question (what is the expected
behaviour of `Delta::iter_deletions`), but it does fix the
crash.

### CODE REVIEW

File: rust/plugin-lib/src/state_cache.rs [hunk 1]
```diff
@@ -521,8 +521,10 @@ impl<'a, S: Default + Clone> PluginCtx<'a, S> {
     /// Updates the line cache to reflect this delta.
     fn update_line_cache(&mut self, delta: &RopeDelta) {
         let (iv, new_len) = delta.summary();
-        if let Some(n) = delta.as_simple_insert() {
-            assert_eq!(iv.size(), 0);
+        // A paste that replaces a selection is treated as a simple insert,
+        // but it screws up our logic, so we pass if the interval size > 0.
```
Comment (vlovich) [inline rust/plugin-lib/src/state_cache.rs]: can you comment specifically how it screws up the logic?

**Comment by vlovich** [comment]
Another comment I'd add is that this code could use some more documentation.  It's not immediately clear what state is being cached & how some of these functions operate.  Might be worth having a little description at the top about what the caching is for (presumably so that plugin can maintain the stateful RPC), how it works (is it caching the document in a local Rope? is it caching other things too?) & how some of these functions modify the cache.  Also identifying which, if any, functions are purely optimizations & the only side-effect of getting it wrong is performance issues then that would be helpful too probably.

**Comment by raphlinus** [comment]
Having read through the deletions iterator I'm pretty sure it's wrong in the case the delta has both insertions and deletions. While this patch seems safe, I think we should fix that.

File: rust/plugin-lib/src/state_cache.rs (hunk 1 above)
Comment (cmyr) [reply rust/plugin-lib/src/state_cache.rs]: Your understanding is correct: the strategy for updating the cache here is that we handle the two common editing cases (typing/deleting) efficiently, and then in the rest of the cases we just blow away the cache.

How this is getting confused is that we were assuming that a 'simple edit' meant an insertion without any deletions, but apparently an insertion that replaces a region was also getting treated as a simple edit. Our logic in `line_cache_simple_insert` is based on the idea that when updating line offsets further down in the cache, the only thing we need to keep track of is the length of the inserted text; but if there's also a deleted region that logic breaks.

**Comment by cmyr** [comment]
@vlovich Agree with the docs comment, opened #527 as a todo for that.

@raphlinus This was my suspicion. I'll clean up my comments in this patch to be a bit clearer, and add a TODO to revisit it once we patch the deletions iter, and then we can merge this as a temporary solution.

**Comment by raphlinus** [comment]
What I think I'd like to do is write a direct is_simple_insert and is_simple_delete on Delta, as these will be more efficient and simpler to understand. That leaves #528 still open, but not on the critical path. I'll prepare a patch shortly.

**Comment by cmyr** [comment]
I actually added `is_simple_delete` and `as_simple_insert` to delta as part of the original patch; they might just need to be touched up?

**Comment by cmyr** [comment]
closing this, #529 resolves

**Pull request closed without merging**